### ISSUE TITLE
Timelock transactions

### ISSUE THREAD

**Issue by limpbrains** [BlueWallet/BlueWallet#1313]
Timelocks are used to lock-up bitcoins held as an investment for a period of months or years.

I don't know any wallet which implements this flawlessly

On the send screen add new option Timelock near the "Allow Fee bump". In this field user can specify the date when this funds will be available. Probably value should be limited to 10-20 years ?

There are a few [options](https://en.bitcoinwiki.org/wiki/Timelock) how to implement this. Need to choose right one

**Comment by junderw** [comment]
NACK without a BIP to explain recovery.

Wallet recovery would be a nightmare if recovering by seed.

**Comment by limpbrains** [comment]
> NACK without a BIP to explain recovery.
> 
> Wallet recovery would be a nightmare if recovering by seed.

Is it because to spend these coins you need to make a transaction with a complex script?

**Comment by junderw** [comment]
1. I want to lock my coins to `2026-11-24T05:53:21.000Z` UTC
2. My redeemScript/witnessScript for my address contains that timestamp (`5126056b`)
3. My phone explodes. All I have is my mnemonic phrase.
4. How should BlueWallet figure out the exact time of `5126056b`?

Some valid answers (not all good UX though) that would need standardizing:

1. Tell the user to remember which key index was used with which timestamp, and remember every timestamp they've ever used.
2. Encode the timestamp data into the mnemonic phrase somehow (this would deviate from BIP39, since there is no way to store arbitrary data) (this would definitely need a BIP, otherwise no other wallet would dare implement it.)
3. Brute force every key + timestamp combination. So for every key you would need to generate an address for every second from 1970 until 2140 (or whenever you run out of seconds)... not feasible.
4. Force users of the timelock wallet to forego the mnemonic backup for a JSON based backup that has all the timestamps included in it.
5. Force users to only use timestamps that are multiples of something. By doing this we reduce the pool of possible timestamps for brute forcing in solution 3. (ie. if they can only choose multiples of 365 days, there would only be 136 possible timestamps to choose from.)

---

tbh none of these really sound like a good idea.

Maybe offering solution 1 as a "you deal with it" situation is fine if we put up tons of warnings.

Basically, the timestamp becomes required info for recovering the coins, so we need to figure out the best way to let the user backup that information.

**Comment by neocogent** [comment]
With CheckTImelockVerify there is no recovery other than waiting. AFAIK. It is possible to use a more complex script to give secondary options. I guess that should be well defined so coders can get it right. 

I wouldn't want to allow 10-20 years. I would say only within next 1-2 years. Users should have to re-lock periodically if they are going to do this because too many unknowns. Angry locked out users would be a big issue.  I think this is why no wallets have really taken on this facility. It sounds cool but there are just too many gotchas and possible situations where legal action could be a problem.

**Comment by junderw** [comment]
> With CheckTImelockVerify there is no recovery other than waiting.

You misunderstand what I am saying.

In my example, even if it was the year 2030 (after the timelock) how will the user remember the exact second they locked their coins in? The only info stored in the blockchain is the P2SH or P2WSH hash. So there's no way to "read" the timestamp from the blockchain. The app or the user needs to remember which key was used with which timestamp.

You need to actually write the timestamp into the script and the timelock / sequence section of the transaction, otherwise, even if the time has passed, you can not spend the coins.

My issue is "what if the user needs to restore a backup? how do we backup the timestamp(s) so that the user can remember the exact timestamps when it comes time to spend?"

**Comment by neocogent** [comment]
The CLTV operator simply compares a parameter against the block height/time (depending on the parameter) and determines if it's satisfied. You can spend any time after that height/time. The user needs to store a redeem script (or be able to regenerate) in order to be able to spend, as otherwise he cannot create the tx, and maybe that's what you meant. Another way is to pre-make a transaction with a future locktime that can be sent (without any additional info) at any time after the set time.

I wrote a plugin for Electtrum to do CLTV called TimeMachine. I got it working pretty nicely except for the spending because I could not find a nice way to integrate into the existing Electrum tx preview/broadcasting process. It could even scan a given range for existing timelocks created with your own keys. It used the same script method as the coinb.in wallet (the only other one I've come across that works). I could use my plugin to lock up small test amounts, and track them in Electrum (it had a tabular timelock list). Anyway, because I could not easily inject the txs for broadcast, and Electrum was changing fast on it's way to 4.0, I just put the whole project aside last year. It's a bit fuzzy in my memory now. It did add an array of time values into the wallet json structure to track the timelocks, and you do need that info in order to regenerate redeem scripts, or you can store redeem scripts.

So I guess the answer is - backup the wallet file. It cannot be done with just a seed. I suppose a time/address list could be stored encrypted on a server. Things like this combined with no great demand are likely why no one's doing it.

If you want to manage it manually, coinb.in does work but is pretty technical to use.

**Comment by limpbrains** [comment]
I agree that without clear recovery process implementing timelocks is not acceptable. Derive timestamp data from BIP39 or bitcoin address would be perfect, I hope that there will be BIP for that someday.

**Comment by Overtorment** [comment]
wild idea: add op_return next to locked output with a hint to recover. 
anyway,  lets keep this ticket open for when there's a proper bip (I see an opportunity here iykwim)

**Comment by neocogent** [comment]
Unfortunately, you need to know the time value to calclulate the relevant address that was paid into as a timelock so you won't know which txs to for a OP_RETURN. You could set a unique value, like a password, but you still need to know that. 

I suppose you could derive it from the seed in a **standardized** way so that every timelock capable wallet knows what value it should search for in the blockchain and then can locate txs which gives it which address is the timelock. But, that does not tell which key was used to create the timelock so it has to work backwards trying every key in sequence to generate a timelock for the exact time of the tx, until the address matches.

Since this involves searching the blockchain it would need to be server based. So a client like Electrum/bluewallet would need to use the seed to generate a value which it passes to the server. The server returns relevant txs data which the client then can trying to match to a key. That actually sounds like it's workable but does rely on having a standard way of generating a value from seed (assuming you want wallet interoperability). Now, someone just needs to write up a BIP...

**Comment by limpbrains** [comment]
If you derive this secret from seed, then you can only create transactions that you send to yourself. It would be nice to send timelock transactions to any address.

**Comment by neocogent** [comment]
You can't do that anyway because a timelock address is based on a script using CLTV op. That means it s P2SH address, ie. a hash of the script (vs. a hash of a multisig or segwit script). It may be possible to write a script that can both receive btc or be a timelock but I'm not aware of how that works. The only ones I've seen (eg. coinb.in has a timelock facility) are purely timelock scripts. They take another p2pkh address (1...) and build the script on that to get a p2sh address. That's something who really knows the scripting inside out could figure out. I'm not fluent enough in those details. 

If someone gives you a timelock address you can still send to it. You just would have to remember it and wouldn't be able to generate it if lost - which is normal even for other address types.

**Comment by limpbrains** [comment]
@neocogent thank you for the explanation!

**Comment by junderw** [comment]
a BIP could be as simple as putting the redeemscript in an OP_RETURN dlrectly after the P2SH/P2WSH output.

you'd be very limited in size of script though.
